timemachine/fe/free_energy.py

```diff
@@ -54,38 +56,37 @@ class SimulationResult:
     dG_errs_png: bytes
     overlap_summary_png: bytes
     overlap_detail_png: bytes
-    frames: List[np.ndarray]
+    frames: List[Sequence[np.ndarray]]  # (len(keep_idxs), n_frames, N, 3)
     boxes: List[np.ndarray]
     initial_states: List[InitialState]
     protocol: SimulationProtocol
 
 
-def image_frames(initial_state: InitialState, frames: np.ndarray, boxes: np.ndarray) -> np.ndarray:
-    """Images a set of frames within the periodic box given an Initial state. Recenters the simulation
-    around the centroid of the coordinates specified by initial_state.ligand_idxs prior to imaging.
+def image_frames(initial_state: InitialState, frames: Sequence[np.ndarray], boxes: np.ndarray) -> np.ndarray:
+    """Images a sequence of frames within the periodic box given an Initial state. Recenters the simulation around the
+    centroid of the coordinates specified by initial_state.ligand_idxs prior to imaging.
 
-    Calling this function on a set of frames will NOT produce identical energies/du_dp/du_dx. Should only
-    be used for visualization convenience.
+    Calling this function on a sequence of frames will NOT produce identical energies/du_dp/du_dx. Should only be used
+    for visualization convenience.
 
     Parameters
     ----------
 
     initial_state: InitialState
         State that the frames came from
 
-    frames: np.ndarray of coordinates
-        Coordinates to image, shape (K, N, 3)
+    frames: sequence of np.ndarray of coordinates
+        Coordinates to image, sequence of K arrays with shape (N, 3)
 
     boxes: list of boxes
-        Boxes to image coordinates into, shape (K, 3, 3)
+        Boxes to image coordinates into, list of K arrays with shape (3, 3)
 
     Returns
     -------
         imaged_coordinates
     """
-    assert len(frames.shape) == 3, "Must be a 3 dimensional set of frames"
-    assert frames.shape[-1] == 3, "Frame coordinates are not 3D"
-    assert boxes.shape[1:] == (3, 3), "Boxes are not 3x3"
+    assert all(frame.ndim == 2 and frame.shape[-1] == 3 for frame in frames), "All frames must have shape (N, 3)"
```

Comment:
If now every time we do `for x in sequence` if `sequence` is backed by disk, we are going to be incurring a lot more cost. Should we move this assertion to the loop?

Reply:
Done in https://github.com/proteneer/timemachine/pull/950/commits/e1d706c82086bca3aeaf1254a812cc3e6168bd3d